### PR TITLE
Hey birt

### PR DESCRIPTION
**Pull request by Jeff-Birt**
I added an 'interrupt controller' which handles all read/write between:
        INTCTRL_START = 0x00_0140;
        INTCTRL_END = 0x00_015F;

This lets the register writes be applied correctly, i.e. to let a register write clear the register. I'm not sure the namespace selection everything is in is the best but I tried to follow the convention of other files.

### CODE REVIEW

**Review by dtremblay — COMMENTED** [review]
Before asking for a pull request, please merge the master branch into your own.

I've tested your code with the Trinity Kernel and several regression errors were introduced, mainly the handling of keyboard and SOF interrupts no longer works.

Also, do not commit the .config file and unused bin files.

**Comment by Jeff-Birt** [comment]
Sorry about that last pull request. For some reason the VS2017 did not push my last commit. I have double checked this one so I hope it is OK.

There were some issues with the memory map and the MemoryManager GetDeviceAt layout the biggest of which was that the CODEC line had the starting address correct but set the length to the same size. I went through the kernel includes and all references I could find to get the memory map as accurate as possible.

Each device that has unique register functionality was moved to its own MemoryRam instance which I think makes it simpler. I think I still need to check out the RAM device postWrite function and reenable it. The interrupt functionality, keyboard, etc. are all functionally separated now.

One side benefit is that I was able to catch any changes (writes) to the LUTs and set a flag. Now the GPU only needs to load the LUTs when there are changes. This saved a little processing time and now I'm seeing up to 3mHz speeds on release builds. I was getting about 2.8mHz before after your last changes.

This push has been merged with 0.2.12.

**Comment by Jeff-Birt** [comment]
Daniel, I'm not sure what I did to offend you but I'm sorry for whatever it was. I'm just trying to help with the project where I can. I have kept the 'Hey_Birt' branch merged with master and just finished merging in your changes from last evening.

I encourage you to have a look. This branch is much faster now. In a release build where debugging data is collect it runs about 5mHz and if you leave the data collection off (i.e. RUN mode) it run about 13.5mHz. A lot of this had to do with not collecting Debug_Lines if the user does not want to (new checkbox on CPU window) or if Debug_Lines are being collected the process is streamlined and no strings are created until things are printed out. This waiting on string creation save a lot of time.

There were some minor changes, in short helper methods that are part of the main CPU execution loop I tried to eliminate intermediate values. This saves some time in in itself and makes it possible for the compiler to decide it can inline a simple method (also faster.) There is probably some of this that can be done with the individaul methods that execute each opcode but I hesitate to do so without a way to test each opcode first.

Also, the memory map is improved and should match in naming and values in the latest kernel. There was one big error in the memeory map which I don't recall off the top of my head but it in in previous commit notes. In the GPU class the LUTs have been changed so they only update if a LUT value is changed in machine memory (i.e. a program writes to the LUT.) This saved some time and I think coudl still be improved maybe by having the Memory_RAM class for Vicky create the LUTs internally and provide a refernece to the GPU class. That way teh Vicky_RAM class can do the updating and the GPU class never has to know about it.

Again, I'm just trying to help out and have some fun learning about the Feonix.

**Comment by dtremblay** [comment]
Hi Jeff,

I have been otherwise engaged with work, so I have been concentrating on
delivering new features for the IDE. Assuming you did a merge of master
into your own branch, merging back to master sound be pretty quick.

I'll try to do this tomorrow. I have to work tonight.

Cheers!

Dan

**Comment by dtremblay** [comment]
OK, a few notes.  I started the merge at 10:30 pm, so I didn't finish.

1. I was trying to get away from the sub-classing of MemoryRAM, so this is
why I've added a "postWrite" method.  It was a mistake on my part to leave
the MathCoproMemoryRam class in.
   Now I see that you've subclassed other devices, like Vicky and Beatrix,
so I would like to bring this back

2. I agree with renaming IO to Vicky, so I'll merge this in.  I also think
that MemoryMap_Other should go.

3. In Operations.cs, you've added this code (lines 955 to 959), I just
don't see the point.  Could you explain why this is there?
        public void ExecuteSTA(byte instruction, AddressModes addressMode,
int signature)
        {




*  if (instruction == 0x9f)            {                int i = 0;
      i++;            }*
            int addr = GetAddress(addressMode, signature, cpu.DataBank);
            cpu.Memory.Write(addr, cpu.A.Value, cpu.A.Width);
        }

I really want to focus on new features.  I'm not too worried about the
performance of the emulator: people who want to run programs in real-time
will need to buy the machine.

I hope this OK.

Cheers!

Daniel Tremblay
Halifax, N.S.

**Comment by Jeff-Birt** [comment]
On #3. Good catch, I should have taken that out. It was just a bit of simple instrumentation to try and catch a problem I was having, which I don't exactly remember what it was. Normally I put a book mark beside a change and then also double check by comparing changed files but I missed that one.

On #1. Here is my train of thought. The MemoryManager class acts like a memory manager/address decode logic does in hardware. In hardware the memory manager makes sure the correct chip is enabled based on the address and in the emulation the MemoryManger class does the same.

In hardware it is apparent that all the different ships, or different parts of a chip (like the SuperIO) are separate entities. In emulation it can be a bit less clear. Many devices will have simple registers that act like reading/writing to RAM so the base MemoryRAM class is just fine, so an instance of the base class is created for it. Other devices like the interrupt controller have very unique requirements, like writing a 1 to a bit in the pending register clears it. The SD card controller, etc. all have other unique requirements. 

Having a subclass of MemoryRAM to handle those features of a device which are unique is sort of the standard way of handling things. You are only overriding those methods that are unique so the additional memory requirements are negligible. It should also lead to better performance as it makes the rest of the code simpler not having to have a lot of separate logic and conditionals to deal with things in multiple parts of an application. The sub class lets you encapsulate everything that is different about a variation into one bit of code making it easier to reason about and maintain. As new features are added it shoudl be simpler to create a subclass of MemoryRAM (if needed) for it and incorporate all the code for it in there.

The other niggling thing I had in the back of my mind with bolting an event driven system on top of the inherently state machine nature of a computer system emulation is that it adds some non-determination to the emulation. In other words when you raise an event (postWrite) when it is handled? The host system will get around to it as soon as it can but the response time is not guarunteed. So, if you have to write to a memory location and then set an IRQ to emulate some device, and the memory write is event driven it might in fact actaully occure after the IRQ is set. If the CPU happens to respond to the IRQ before the corresponding memory location was written to what would happen? This could lead to some very interesting bugs.

I've not done a lot with coding computer system emualtions but I do have a good deal of experiance with machine control applciations (PC based, PLC based, etc.) This leads me to always being very cautious about 'when' things happen and what state the phyical hardware will be in. It is just a differnet point of view.

In short I think subclassing the individual devices is the best way to go. Overall it will make things simpler and allow for consolidating all code related to a device in one area. I'm not sure what possible detrimetn there might be to having sepeate subcalsses?

**Comment by dtremblay** [comment]
I started the merge.  I've removed the delegate method to use subclasses instead. I agree that it makes the code more readable.  I need to cherry-pick the changes you've made because I want to maintain and extend the interrupt handling eventually.  The current code base handles SOF interrupts quite nicely and it doesn't slow down the system in any way.  It reproduces the actual system more accurately.

**Comment by Jeff-Birt** [comment]
Out of curiosity what is different about how SOF is handled now as opposed to the HW? Just wanting to learn more about the actual hardware..

**Comment by dtremblay** [comment]
The SOF in the hardware is raised before the first display scan line is drawn.
We are also doing this in the emulator, although it's done before we start drawing anything.

The SOL (start of line) interrupt is raise midway in the screen rendering.  The emulator is not doing anything so far, because we are not drawing the screen the same way Vicky is.

We may decide to change how our GPU renders by lines, instead of drawing a full 640x480 bitmap for each layer (tiles, sprites, text).

I also want to start handling the mouse via interrupts as well, as opposed to each time the GPU display is rendered.  The reason for this is that the emulator is not dealing with mouse clicks and moves via interrupts, so the assembly code doesn't get exercised.

**Comment by dtremblay** [comment]
Hi Jeff,
I've had to do a manual merge... yes, I'm picky too.  I've applied most of your changes but not necessarily in the same structure.  I hope this is OK with you: your improvements are helping.

So, what I would like you to do now, is start a fresh branch and apply the changes that you really care about, such as the CPU pins, for example.  Then I'll just merge this in.

Keeping your merge requests small will make the process easier too.

So, I know we can eke out more performance, and most of this will have to come from the drawing function.  You can look there if you wish.
